src/local-mode.ts

```diff
@@ -56,51 +102,159 @@ export class LocalModeManager {
 
 		if (!localMode) {
 			// Remove any leftovers if necessary
-			await this.removeLocalModeArtifacts();
+			await this.handleLocalModeStateChange(false);
 		}
 	}
 
-	public async removeLocalModeArtifacts(): Promise<void> {
-		try {
-			const images = await this.getLocalModeImages();
-			const containers = await this.getLocalModeContainers(images);
-
-			await Bluebird.map(containers, containerId => {
-				log.debug('Removing local mode container: ', containerId);
-				return this.docker.getContainer(containerId).remove({ force: true });
-			});
-			await Bluebird.map(images, imageId => {
-				log.debug('Removing local mode image: ', imageId);
-				return this.docker.getImage(imageId).remove({ force: true });
-			});
-
-			// Remove any local mode state added to the database
-			await this.db
-				.models('app')
-				.del()
-				.where({ source: 'local' });
-		} catch (e) {
-			log.error('There was an error clearing local mode artifacts: ', e);
-		}
+	public startLocalModeChangeHandling(local: boolean) {
+		this.switchInProgress = new Promise(async resolve => {
+			try {
```

Comment:
This `try...finally` makes me a little uneasy, if the `handleLocalModeStateChange(..)` function ever changes to throw an error (I know currently it can't) we'd be ignoring it and printing nothing. I think given that we know it cannot fail, I'd prefer this to be removed, and in the case where the `handleLocalModeStateChange` function changes to ever throw an error, we add this back in with an explicit `catch`.

I'm happy to hear if you have any other thoughts on it though.

Reply:
`startLocalModeChangeHandling` is responsible for managing `switchInProgress` variable, and I was trying to make it as independent from the rest of the logic as possible.
Its responsibility is to construct a promise which gets resolved whenever local mode switch handler completes regardless how it completes to avoid any possibility of a deadlock. That is why, I put a `finally` block inhere.

In fact, it can be written "simpler", I think:
```
this.switchInProgress = this.handleLocalModeStateChange(local)
    .finally(() => { this.switchInProgress = null; });
```
But it does not compile with our current setup :) TS compiler says there is no `finally` property in `Promise`.

Does it make sense? It's not really about handling errors, but about separating the task of unblocking other operations whenever local mode switch completes.

Reply:
Side note: this method should actually be private.
But I need to use it in tests... Do you know a good way how I could limit the method visibility to tests only?

Should I look into namespaces?

Reply:
I don't think there's a good solution to this, it's either make it public, or do `(instance as any).method` in the tests, losing types.

Reply:
I'm pretty sure the method you mention here https://github.com/balena-io/balena-supervisor/pull/1012#discussion_r297566134 will propagate errors so I'd be fine with this. To do that, you'd have to wrap it in a `Bluebird.resolve`.

Reply:
Done: https://github.com/balena-io/balena-supervisor/pull/1012/files#diff-0f5fedf56e0f6f78a6c52a828857e5e4R110

Reply:
So do we need to use Bluebird because of older nodejs versions?

Reply:
Actually, I forgot that `.finally` isn't a bluebird extension, and should be present in the native. I'd check if we need to upgrade our `@types/node`.

Reply:
Ok, resolving this thread in such case.